### ISSUE TITLE
Feature Request: explorer integration for VS2019

### ISSUE THREAD

**Issue by benblo** [oleg-shilo/cs-script#228]
Hi! I use the Windows Explorer integration a lot, and recently migrated to VS2019. So I also upgraded cs-script (through Chocolatey), but the shell-x shortcuts don't contain anything for 2019. 
I vaguely tried copy-pasting debug15.0.cs to debug16.0.cs and replace a few version numbers, but then I noticed it's pointing to a lib/Debug/VS15.0 folder, and there is no VS16 version of that, so I'm not sure my blind attempt is very productive unless I know what I'm doing :).
I think in the past few years Microsoft changed the csproj format (?), so I'm not sure how compatible 2017 and 2019 are, but since your VSIX plugin also supports 2019, it obviously can handle 2019 csprojs: could it also handle explorer integration?

Somewhere along the way (during installation of VS 2019? or CS-Tools 2019 maybe?) the VS 2017 integration actually broke as well! Any idea how to fix that?
![image](https://user-images.githubusercontent.com/574187/107917101-ad9d1680-6f67-11eb-9bca-7086374446c5.png)

I really like CS-Script, being able to script in C# (instead of Python or something) is invaluable in companies where the main codebase is already C#... but overall the install experience is really awkward IMO: Chocolatey install is great, then run css_config (meh, could be part of choco install); then you also have to install shell-x, and copy-paste a folder in AppData... meh.
What makes this more difficult is the documentation and examples are full of obsolete information, so it's hard to tell if I'm doing something wrong or not:

- some examples mention css.exe, the debugVS15.0.cs help mentions "cscscript", none of which exist in my install
- mentions and screenshots of non-shell-x integration
- mentions of C# 6 and setting up a custom Roslyn, which I'm not sure is even necessary anymore
- lots of dead links
- the new [explorer integration documentation](https://github.com/oleg-shilo/cs-script/wiki/REPL,-IDE-and-OS-integration) is a bit buried, no link to it from the main readme
- the [cs-script.net](https://www.cs-script.net/) website is _very_ obsolete

I've been thinking for a while of deploying cs-script company-wide to use as the pipeline backbone, but I'm reluctant because it feels brittle. I wish for a "state of the art in 2021" subset of cs-script, that "just works" in a single step, and with all the old extra stuff removed.

Oh and last thing, the "Recent Scripts" section of the VS plugin is broken, and if you click "Manage Recent Scripts" it opens VS2012_Recent.txt, so it looks like it's pointing to the wrong version.

Overall it feels like the VS plugin is lagging behind a bit, I noticed the VsCode plugin seems more recent: did you switch your focus to VSCode? (I wouldn't blame you, VSCode is awesome :).

I'm sorry for what probably looks like a whiny report, and a bit of a kitchen sink too; let me know if you want me to break it up in several issues.

**Comment by oleg-shilo** [comment]
I have just released CS-Script.Core v.2.0.0.0 for .NET5. (Choco is yet to be published)

https://github.com/oleg-shilo/cs-script.core/releases/tag/v2.0.0.0

And this release is to be re-released as CS-Script v.4.0.0.0 within a few days. From that moment it will be no more CS-Script.Core but CS-Script only. 

After that I will be able to start addressing the gaps in the documentation and the project website.

> "state of the art in 2021" subset of cs-script, that "just works" 

That's exactly why CS-Script.Core stream was intended and with the release of .NET5 it becomes the only stream supported.
This edition is much lighter and the old legacy approaches have been made obsolete. 
The trimming old no longer valuable features has started long time ago with CS-Script.Core ago I was avoiding over-investing in the morally dead .NET Framework. Thus you will find that there is no "installation" any more, no need for Roslyn, very small distro.  
And now with the introduction of .NET5 I am able to promote CS-Script.Core into the only CS-Script supported.

You will benefit from the new release because handling VS is embedded in the script engine and no extra infrastructure is required.
Thus opening a script in VS only takes this as `open with VS2019.cmd`:
```ps
css -vs myscript.cs
```

I do not mark this issue as fixed/addressed as the release is yet to be published outside of GitHub and the documentation is yet to be cleaned.

**Comment by oleg-shilo** [comment]
just published on choco

**Comment by benblo** [comment]
This all sounds awesome! Modern .NET Core, embedded VS handling, all great stuff 👍 !

Unfortunately, so far I'm unable to use it:

- got a virus warning during the install via chocolatey :
```powershell
PS C:\WINDOWS\system32> choco install cs-script.core
Chocolatey v0.10.15
Installing the following packages:
cs-script.core
By installing you accept licenses for the packages.
Progress: Downloading cs-script.core 2.0.0.0... 100%
cs-script.core not installed. An error occurred during installation:
 Operation did not complete successfully because the file contains a virus or potentially unwanted software.

cs-script.core package files install completed. Performing other installation steps.
The install of cs-script.core was NOT successful.
cs-script.core not installed. An error occurred during installation:
 Operation did not complete successfully because the file contains a virus or potentially unwanted software.


Chocolatey installed 0/1 packages. 1 packages failed.
 See the log for details (C:\ProgramData\chocolatey\logs\chocolatey.log).

Failures
 - cs-script.core (exited 1) - cs-script.core not installed. An error occurred during installation:
 Operation did not complete successfully because the file contains a virus or potentially unwanted software.

```
(This is from a second attempt, after it had already installed the .NET Core dependency, then failed on cs-script.core.)
- So after fiddling in Windows Defender I found this:
![image](https://user-images.githubusercontent.com/574187/109501164-83834280-7a97-11eb-9887-08927f9f175d.png)
... and allowed the "threat": install now went through successfully (screencap from chocolatey-gui):
![image](https://user-images.githubusercontent.com/574187/109502226-02c54600-7a99-11eb-96ef-27f798084721.png)

- I opened a terminal, it found `css` instantly (without having to logoff or restart or refreshenv or whatever: yay!)... unfortunately:
```powershell
PS C:\Users\Benoit FOULETIER> css
It was not possible to find any compatible framework version
The framework 'Microsoft.NETCore.App', version '5.0.0' was not found.
  - The following frameworks were found:
      2.0.5 at [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
      2.0.7 at [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
      2.0.9 at [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
      2.1.5 at [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
      2.1.24 at [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
      3.0.1 at [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
      3.1.7 at [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
      3.1.12 at [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]

You can resolve the problem by installing the specified framework and/or SDK.

The specified framework can be found at:
  - https://aka.ms/dotnet-core-applaunch?framework=Microsoft.NETCore.App&framework_version=5.0.0&arch=x64&rid=win10-x64
```
Same result with both cmd and powershell.

I'm gonna try rebooting, or downloading said framework from the link (although I think it should be covered by the choco dependency, shouldn't it? or it only an sdk, not the runtime?)

**Comment by benblo** [comment]
So, rebooting didn't work, but installing the framework from the aka.ms link did:
```powershell
> css
C# Script execution engine (.NET Core). Version 2.0.0.0.
Copyright (C) 2004-2020 Oleg Shilo. www.csscript.net (github.com/oleg-shilo/cs-script.core)

   CLR:             5.0.3
   System:          Microsoft Windows NT 10.0.19042.0
   Architecture:    x64
   Install dir:     C:\ProgramData\chocolatey\lib\cs-script.core\tools
   Script engine:   C:\ProgramData\chocolatey\lib\cs-script.core\tools\cscs.dll
   Config file:     <none>
   Compiler engine: dotnet (C:\Program Files\dotnet\dotnet.exe)
   NuGet manager:   dotnet
   NuGet cache:     C:\Users\Benoit FOULETIER\.nuget\packages
   Custom commands: C:\ProgramData\cs-script\commands
   Global includes: C:\ProgramData\CS-Script\inc
```
... I'm in the Matrix!

---
However, when trying the new syntax to open in vs:
```powershell
> css -vs test.cs
Opening project: C:\Users\Benoit FOULETIER\AppData\Local\Temp\csscript.core\.vs\-1684661676\.build\test.cs\test.csproj
Error: you need to set environment variable 'CSSCRIPT_VSEXE' to the valid path to Visual Studio executable devenv.exe.
```
No biggie, I'll set the env var, but it would be even nicer if it was auto-detected on install or something.
(Also, may I ask: why use an env var and not a setting in the config file?)

Env var set, running again:
![image](https://user-images.githubusercontent.com/574187/109506772-538b6d80-7a9e-11eb-952a-eea25274147f.png)
I have a space in my user name, that's probably the issue?

---
OK, trying VsCode now:
```powershell
> css -vscode test.cs
Opening script: D:\test.cs
Error: you need to set environment variable 'CSSCRIPT_VSCODEEXE' to the valid path to Visual Studio Code executable codecode.exe.
```
Same issue, setting the var (btw, the exe name is just `code.exe`, not `codecode.exe` - for me: `C:\Program Files (x86)\Microsoft VS Code\Code.exe`).
VsCode loads ok this time, but when I run (with the |> button in the toolbar): 
![image](https://user-images.githubusercontent.com/574187/109507836-6f434380-7a9f-11eb-8869-673810261258.png)

```
[Running] dotnet "C:\Users\Benoit FOULETIER\AppData\Roaming\Code\User\cs-script.user\dotnet\cscs.dll" "d:\test.cs"
Error: Specified file could not be compiled.

d:\test.cs(0,0): error CS1705: Assembly 'cscs' with identity 'cscs, Version=1.4.5.0, Culture=neutral, PublicKeyToken=null' uses 'System.Runtime, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' which has a higher version than referenced assembly 'System.Runtime' with identity 'System.Runtime, Version=4.2.2.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' [C:\Users\Benoit FOULETIER\AppData\Local\Temp\csscript.core\cache\769641162\.build\test.cs\test.csproj]
C:\Program Files\dotnet\sdk\3.1.406\Microsoft.Common.CurrentVersion.targets(2084,5): warning MSB3277:  Found conflicts between different versions of "System.Collections" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed. 
C:\Program Files\dotnet\sdk\3.1.406\Microsoft.Common.CurrentVersion.targets(2084,5): warning MSB3277:  Found conflicts between different versions of "System.ComponentModel.Primitives" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed. 
C:\Program Files\dotnet\sdk\3.1.406\Microsoft.Common.CurrentVersion.targets(2084,5): warning MSB3277:  Found conflicts between different versions of "System.ComponentModel.TypeConverter" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed. 
C:\Program Files\dotnet\sdk\3.1.406\Microsoft.Common.CurrentVersion.targets(2084,5): warning MSB3277:  Found conflicts between different versions of "System.Console" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed. 
C:\Program Files\dotnet\sdk\3.1.406\Microsoft.Common.CurrentVersion.targets(2084,5): warning MSB3277:  Found conflicts between different versions of "System.Diagnostics.Process" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed. 
C:\Program Files\dotnet\sdk\3.1.406\Microsoft.Common.CurrentVersion.targets(2084,5): warning MSB3277:  Found conflicts between different versions of "System.Diagnostics.TraceSource" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed. 
C:\Program Files\dotnet\sdk\3.1.406\Microsoft.Common.CurrentVersion.targets(2084,5): warning MSB3277:  Found conflicts between different versions of "System.IO.FileSystem.Watcher" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed. 
C:\Program Files\dotnet\sdk\3.1.406\Microsoft.Common.CurrentVersion.targets(2084,5): warning MSB3277:  Found conflicts between different versions of "System.IO.FileSystem" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed. 
C:\Program Files\dotnet\sdk\3.1.406\Microsoft.Common.CurrentVersion.targets(2084,5): warning MSB3277:  Found conflicts between different versions of "System.Linq.Expressions" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed. 
C:\Program Files\dotnet\sdk\3.1.406\Microsoft.Common.CurrentVersion.targets(2084,5): warning MSB3277:  Found conflicts between different versions of "System.Linq" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed. 
C:\Program Files\dotnet\sdk\3.1.406\Microsoft.Common.CurrentVersion.targets(2084,5): warning MSB3277:  Found conflicts between different versions of "System.Net.Primitives" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed. 
C:\Program Files\dotnet\sdk\3.1.406\Microsoft.Common.CurrentVersion.targets(2084,5): warning MSB3277:  Found conflicts between different versions of "System.Net.Sockets" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed. 
C:\Program Files\dotnet\sdk\3.1.406\Microsoft.Common.CurrentVersion.targets(2084,5): warning MSB3277:  Found conflicts between different versions of "System.Runtime.InteropServices" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed. 
C:\Program Files\dotnet\sdk\3.1.406\Microsoft.Common.CurrentVersion.targets(2084,5): warning MSB3277:  Found conflicts between different versions of "System.Runtime" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed. 
C:\Program Files\dotnet\sdk\3.1.406\Microsoft.Common.CurrentVersion.targets(2084,5): warning MSB3277:  Found conflicts between different versions of "System.Text.RegularExpressions" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed. 
C:\Program Files\dotnet\sdk\3.1.406\Microsoft.Common.CurrentVersion.targets(2084,5): warning MSB3277:  Found conflicts between different versions of "System.Threading.Thread" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed. 
C:\Program Files\dotnet\sdk\3.1.406\Microsoft.Common.CurrentVersion.targets(2084,5): warning MSB3277:  Found conflicts between different versions of "System.Threading" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed. 
C:\Program Files\dotnet\sdk\3.1.406\Microsoft.Common.CurrentVersion.targets(2084,5): warning MSB3277:  Found conflicts between different versions of "System.Xml.ReaderWriter" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed. 
C:\Program Files\dotnet\sdk\3.1.406\Microsoft.Common.CurrentVersion.targets(2084,5): warning MSB3277:  Found conflicts between different versions of "System.Xml.XDocument" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed. 
C:\Program Files\dotnet\sdk\3.1.406\Microsoft.Common.CurrentVersion.targets(2084,5): warning MSB3277:  Found conflicts between different versions of "System.Xml.XmlSerializer" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed. 
d:\test.cs(0,0): error CS1705: Assembly 'cscs' with identity 'cscs, Version=1.4.5.0, Culture=neutral, PublicKeyToken=null' uses 'System.Runtime, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' which has a higher version than referenced assembly 'System.Runtime' with identity 'System.Runtime, Version=4.2.2.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' [C:\Users\Benoit FOULETIER\AppData\Local\Temp\csscript.core\cache\769641162\.build\test.cs\test.csproj]


[Done] exited with code=0 in 1.985 seconds

```
That's with a script freshly created with `css -new test.cs`, btw.
After fiddling around, if I use `Console.WriteLine` instead of `dbg.print` (which I actually never heard about before this), it works! Victory!
Is `dbg` part of some newfangled .NET Core assembly I don't have locally? `Console.WriteLine(Environment.Version);` outputs 5.0.3, as expected.

---
From that super-brief experience though, it seems to run quite slower on first run/compile, right? I sorta remember reading somewhere that it was to be expected...?
Like, I added a simple output at the top of the main function:
```cs
		Console.WriteLine("hello console");
		//dbg.print("hello print");
		(string message, int version) setup_say_hello()
		{
			return ("Hello from C#", 9);
		}

		var info = setup_say_hello();
		//print(info.message, info.version);
		Console.WriteLine(info.message);
```
it actually displays "hello console", stalls, _then_ displays "hello from C#"... is it compiling the nested function in-between?

---
I also tried the "run in terminal" button ![image](https://user-images.githubusercontent.com/574187/109509167-e7f6cf80-7aa0-11eb-9140-e8f402d08cb6.png), it briefly opens a terminal, displays a few lines in red (never a good sign 😄) that I don't have time to read before it disappears. If I press the button again, nothing happens.

**Comment by oleg-shilo** [comment]
Yes the presence of .NET5 framework is a must.
I could not introduce the .NET5 dependency in the choco package the first pre-release as the .NET was not fully out at that time.

but now it is time to reestablish this dependency. Will update the package definition.

**Comment by oleg-shilo** [comment]
> run quite slower on first run/compile, right?

yep, that' the usability issue .NET5 and .NET Core have. Not much we can do about it. At least cs-script caches the compilers and you can run very fast next time if no code changed.

You can also improve improve the speed even if cases where code is changed if you us cs-script build server instead of raw `dotnet.exe`. It uses the same approach as early Rolsyn did with VBCSCompiler.exe.

Place on the top of your script this directive:
```C#
//css_ng csc
```

and it will use cached csc.exe compiler that is much faster than dotnet.exe

As fro the code you are asking about:
```C#
01. Console.WriteLine("hello console");
02. //dbg.print("hello print");
03. (string message, int version) setup_say_hello()
04. {
05.    return ("Hello from C#", 9);
06. }
07.
08. var info = setup_say_hello();
09. //print(info.message, info.version);
10. Console.WriteLine(info.message);
```
>  is it compiling the nested function in-between?

Yes. This is a normal C# behavior. Well this syntax is C#9. 
Line `#1` prints the first message and line `#10` the second one

As for terminal. It's possibly something related to VSC itself. The only thing that this button does is 
1. make terminal window visible
2. execute `dotnet <script_engine_file> <file_opened_in_editor>`

![image](https://user-images.githubusercontent.com/16729806/109598656-cca5c600-7b6d-11eb-8d65-9449bcec0db5.png)

**Comment by oleg-shilo** [comment]
> I also noticed that I did not answer all your questions.

In the latest release v2.0.0.0 both VS abd VSCode are auto-dicovered. And if the dicovery fails then it falls back to envar.
I opted to envar for not strong reason. I consider -vs and -vscode` as still experimental so wanted not po polut config too much. Though if auto-discovery is plays nice then even less reasons to do it in config.

As for `print()` it is an extension method that is available from the cscs.dll (the script engine) that all script reference. 
I have checked the VSCode and it seems to work:
```C#
//css_ng csc
using System;

Console.WriteLine(111);
dbg.print(222);
3333.print();
user().print();

string user()
    => Environment.UserName;
```
![image](https://user-images.githubusercontent.com/16729806/110208129-6c9b8080-7edb-11eb-8015-c9177a6b0f3e.png)

Not sure why it does not in your case.
Not handling space in the path is a mistake.  This is how cs-scripts opens th project:

```
p = Process.Start(vs_exe, projectFile);
```

Will be fixed in the next release
